### PR TITLE
FP S5332: Improve detection of namespace uris

### LINKED ISSUE #6223: FP S5332: Improve detection of namespace uris
Uris used for namespaces start with `http` and are not reported by S5332. The detection of namespace uris can be improved:

Copied from 
https://github.com/SonarSource/sonar-dotnet/issues/6141#issuecomment-1277250021

* [x] [XmlnsDictionary](https://learn.microsoft.com/dotnet/api/system.windows.markup.xmlnsdictionary)
* [x] [XmlSerializerNamespaces](https://learn.microsoft.com/dotnet/api/system.xml.serialization.xmlserializernamespaces)
* [x] [XmlQualifiedName construktor](https://learn.microsoft.com/de-de/dotnet/api/system.xml.xmlqualifiedname.-ctor?view=net-6.0#system-xml-xmlqualifiedname-ctor(system-string-system-string))
* [x] [XmlNamespaceManager Class (System.Xml)](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlnamespacemanager?view=net-6.0)
* [x] [XmlAttribute(String, String, String, XmlDocument) Constructor (System.Xml) | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlattribute.-ctor?view=net-6.0#system-xml-xmlattribute-ctor(system-string-system-string-system-string-system-xml-xmldocument))
* [x] [XmlElement(String, String, String, XmlDocument) Constructor (System.Xml) | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlelement.-ctor?view=net-6.0#system-xml-xmlelement-ctor(system-string-system-string-system-string-system-xml-xmldocument))
* [x] [XmlWriter.WriteStartElement Method (System.Xml)](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlwriter.writestartelement?view=net-6.0#system-xml-xmlwriter-writestartelement(system-string-system-string))
* [x] [XmlWriter.WriteElementString Method (System.Xml) | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlwriter.writeelementstring?view=net-6.0#system-xml-xmlwriter-writeelementstring(system-string-system-string-system-string))
* [x] [XmlSerializationWriter.WriteElementStringRaw Method (System.Xml.Serialization) | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializationwriter.writeelementstringraw?view=netcore-3.1)

Update list of well-known namespaces
* [x] https://referencesource.microsoft.com/#system.xml/System/Xml/XmlReservedNs.cs
* [x] Add oasis-open.org namespaces (e.g. wss, see e.g. [peach](https://peach.sonarsource.com/security_hotspots?id=wcf&hotspots=AXddXtF0Po2w9IEVshSJ)):
http://docs.oasis-open.org/wss/2004/01/
* [x] Add ws-i.org namespaces http://ws-i.org/ e.g. [peach](https://peach.sonarsource.com/security_hotspots?id=wcf&hotspots=AXZ4YUlp7R5umuxffeXo)

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #6223

### CODE REVIEW

**Review by Tim-Pohlmann — APPROVED** [review]
The code looks fine, but I can't fully understand the context of the rule. Maybe get another review.
I added one optional code refactoring.

File: analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            if (symbol is null || symbol.Name != ns[i])
+            {
+                return i == 0 && ns[i] == "global" && symbol?.IsGlobalNamespace is true;
+            }
+            else
+            {
+                symbol = symbol.ContainingNamespace;
+            }
+        }
+        return symbol is null || symbol.IsGlobalNamespace;
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs]: Nice, but I removed the "global" handling instead. The loop is much easier to read now and should be fine (it also allocates a lot less and should be straight forward to read now).

**Comment by martin-strecker-sonarsource** [comment]
@pavel-mikula-sonarsource Can you please merge. The missing coverage isn't really a concern IMO.

File: analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs:21]:
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+internal static class INamespaceSymbolExtensions
+{
+    /// <summary>
+    /// Checks if the <see cref="INamespaceSymbol"/> fits the <paramref name="name"/>. The format of <paramref name="name"/> is the same as in a <see langword="using"/> directive.
+    /// </summary>
+    /// <param name="symbol">The namespace symbol to test.</param>
+    /// <param name="name">The name in the form <c>System.Collections.Generic</c>.</param>
+    /// <returns>Returns <see langword="true"/> if the namespace symbol refers to the string given.</returns>
+    public static bool Is(this INamespaceSymbol? symbol, string name)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs:31]:
```suggestion
    public static bool Is(this INamespaceSymbol symbol, string name)
```

File: analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public static bool Is(this INamespaceSymbol symbol, string name)
+    {
+        if (name is null)
+        {
+            throw new ArgumentNullException(nameof(name));
+        }
+        if (symbol is null)
+        {
+            return false;
+        }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/INamespaceSymbolExtensions.cs:40]: This is the same thing with way less code, isn't it? The logic is null tolerant and would return `false` by itself.
```suggestion
        _ = name ?? throw new ArgumentNullException(nameof(name));
```

**Comment by martin-strecker-sonarsource** [comment]
Validation: The AD001 is gone.

**Comment by martin-strecker-sonarsource** [comment]
Validation:
Lost FPs:
* AWSSDK: 760 calls like this `xmlWriter.WriteStartElement("ForwardedValues", "http://cloudfront.amazonaws.com/doc/2018-11-05/")`
* dnn: 3 due to `writer.WriteStartElement` and `http://www.sitemaps.org` namespace
* dotnet-runtime: 2 because of inclusion of the URIs defined in `XmlReservedNs`
* odata: 1 "http://www.opengis.net"
* ravendb: 1 because of `XNamespace.Get`
* smartstore: 2 because of `XmlNamespaceManager` support
* wcf: 277 because of the addition of several namespace uris.

Add TPs:
* ocelot: 2 new issues. Reason unknown.

**Pull request merged**